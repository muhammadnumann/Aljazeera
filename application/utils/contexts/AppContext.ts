import { createContext } from 'react'

export interface ContextProps {
  serverUrl: string | undefined
  username: string | undefined
  token: string | undefined
  userPermissions: string | string[]
  environment: string | undefined
}

const AppContext = createContext({} as ContextProps)

export default AppContext
