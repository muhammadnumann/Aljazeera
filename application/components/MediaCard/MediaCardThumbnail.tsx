import React, {
  MouseEvent,
  MouseEventHandler,
  useContext,
  useRef,
  useState,
} from 'react'
import AppContext from 'utils/contexts/AppContext'
import { ThumbnailType } from '../../types'

interface InputProps {
  thumbnail: string
  itemId?: string
  onClick?: MouseEventHandler
  spriteImage?: HTMLImageElement
  thumbnailArray?: Array<ThumbnailType>
}

const MediaCardThumbnail = (props: InputProps) => {
  const context = useContext(AppContext)
  const [scrubPosition, setScrubPosition] = useState(0)
  const [showScrub, setShowScrub] = useState(false)
  const [srcThumbnail, setSrcThumbnail] = useState(
    `${context.serverUrl}${props.thumbnail}`,
  )
  const [x, setX] = useState(0)
  const [y, setY] = useState(0)

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const divThumbnail = useRef<any>()

  const { itemId, spriteImage, thumbnailArray } = props

  const getOffsetLeft = (element: HTMLElement) => {
    try {
      return Math.round(element.getBoundingClientRect().left)
    } catch (error) {
      return 0
    }
  }

  const getMouseElementPosition = (e: MouseEvent<HTMLImageElement>) => {
    const mouseElementPosition = e.clientX - getOffsetLeft(e.currentTarget)
    const elementWidth = e.currentTarget!.offsetWidth
    if (mouseElementPosition < 0) {
      return 0
    }
    if (mouseElementPosition > elementWidth) {
      return elementWidth
    }
    return mouseElementPosition
  }

  const onMouseMove = (e: MouseEvent<HTMLImageElement>) => {
    if (spriteImage && spriteImage.complete) {
      console.info('spriteImage.complete')
      const mouseElementPosition = getMouseElementPosition(e)
      /*
       *   Calculate the thumbnail based on the mouse position
       *   index = (mouse position / ( element width / number of thumbnails))
       */
      // @ts-ignore
      const index = Math.floor(
        mouseElementPosition /
          ((divThumbnail.current ? divThumbnail.current.offsetWidth : 0) /
            thumbnailArray!.length || 1),
      )
      if (!!thumbnailArray && index >= 0 && index < thumbnailArray.length) {
        setX(thumbnailArray[index].x)
        setY(thumbnailArray[index].y)
      }
      setScrubPosition(mouseElementPosition - 1)
    }
  }

  const handleShowScrub = (value: boolean) => () => {
    if (spriteImage && spriteImage.complete) {
      setSrcThumbnail(spriteImage.src)
      setShowScrub(value)
    }
  }

  return (
    <div
      className="flex flex-col items-center justify-center cursor-pointer"
      onMouseEnter={handleShowScrub(true)}
      onMouseLeave={handleShowScrub(false)}
      onMouseMove={onMouseMove}
      onClick={props.onClick}
    >
      <div className="flex flex-col items-center justify-center relative">
        <div
          ref={divThumbnail}
          className={
            'rounded-t-lg w-[256px] h-[140px] overflow-hidden relative'
          }
        >
          <img
            alt={`${itemId} thumbnail`}
            className={`rounded-t-lg max-w-none absolute ${
              spriteImage?.complete ? 'invisible' : 'visible'
            }`}
            src={`${context.serverUrl}${props.thumbnail}`}
          />
          <img
            alt={`${itemId} thumbnail`}
            className={`rounded-t-lg max-w-none absolute ${
              spriteImage?.complete ? 'visible' : 'invisible'
            }`}
            style={{
              top: `-${y}px`,
              left: `-${x}px`,
            }}
            src={`${srcThumbnail}`}
          />
        </div>
        {showScrub && (
          <div
            className="absolute h-full w-[2px] bg-gray-200"
            style={{
              left: scrubPosition,
            }}
          />
        )}
      </div>
    </div>
  )
}

export default MediaCardThumbnail
