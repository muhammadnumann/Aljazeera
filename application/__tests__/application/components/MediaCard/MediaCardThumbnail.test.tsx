import React from 'react'
import AppContext from '../../../../utils/contexts/AppContext'
import MediaCardThumbnail from '../../../../components/MediaCard/MediaCardThumbnail'
import toJson from 'enzyme-to-json'
import { shallow } from 'enzyme'
import { DUMMY_USERNAME, DUMMY_SERVER_URL, DUMMY_TOKEN } from '../../../helpers/fixture.helper'

describe('<MediaCardThumbnail />', () => {

    it('should render correctly', () => {

        const component = shallow(
        <AppContext.Provider value={{ username: DUMMY_USERNAME, serverUrl: DUMMY_SERVER_URL, token: DUMMY_TOKEN }}>;
            <MediaCardThumbnail thumbnail='/path/to/img.jpg' itemId='VX-1' />);
        </AppContext.Provider>);
            
        expect(toJson(component)).toMatchSnapshot();
    });
});
