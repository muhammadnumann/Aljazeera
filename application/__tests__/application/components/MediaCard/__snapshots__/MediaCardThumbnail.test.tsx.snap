// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MediaCardThumbnail /> should render correctly 1`] = `
[
  ";",
  <MediaCardThumbnail
    itemId="VX-1"
    thumbnail="/path/to/img.jpg"
  />,
  ");",
]
`;
